addons/mail/static/src/discuss/call/common/call_settings.js

```diff
@@ -65,6 +86,39 @@ export class CallSettings extends Component {
         return isMobileOS();
     }
 
+    async testMicrophoneVolume() {
+        if (!this.state.isTestMicrophoneReady) {
+            return;
+        }
+        this.state.isTestMicrophoneReady = false;
+        this.disconnectAudioMonitor?.();
+        this.disconnectAudioMonitor = undefined;
+        if (this.audioTrack) {
+            this.audioTrack.stop();
+            this.audioTrack = null;
+            this.state.isTestMicrophoneReady = true;
+            this.state.volume = 0;
+            return;
+        }
+        const audioStream = await browser.navigator.mediaDevices.getUserMedia({
+            audio: this.store.settings.audioConstraints,
+        });
+        const track = audioStream.getAudioTracks()[0];
+        if (this.isClosed) {
+            track.stop();
+            return;
+        }
+        this.audioMonitorPromise = monitorAudio(track, {
```

Comment:
I feel that all these audio stuffs could be put in a hook, so that it is much more reusable and it doesn't pollute a component with many concerns with lots of code related to a very specific feature in it. This is the call settings menu, and a big chunk of attributes and code come from the testing of voice detection, which in practice is not that frequent. The component would be simpler to maintain if this was in a hook.